README.md
# node-bswap

x86: ![x86 Build Status](https://github.com/zbjornson/node-bswap/actions/workflows/ci.yaml/badge.svg)  
ARM: [![ARM Build Status](https://cloud.drone.io/api/badges/zbjornson/node-bswap/status.svg)](https://cloud.drone.io/zbjornson/node-bswap)

The fastest function to swap bytes (a.k.a. reverse the byte ordering, change
endianness) of TypedArrays in-place for Node.js and browsers. Uses SIMD when
available. Works with all of the TypedArray types, including BigUint64Array and
BigInt64Array. Also works on Buffers if you construct a TypedArray view on the
underlying ArrayBuffer (see below).

Install:
```
$ npm install bswap
```

Use:
```js
import bswap from "bswap";
const x = new Uint16Array([1, 2, 3, 4, 5, 6, 7, 8]);
bswap(x);
// now: Uint16Array [ 256, 512, 768, 1024, 1280, 1536, 1792, 2048 ]

// With buffers:
const b = Buffer.alloc(128);
// This constructs a "view" on the same memory; it does not allocate new memory:
const ui32 = new Uint32Array(b.buffer, b.byteOffset, b.byteLength / Uint32Array.BYTES_PER_ELEMENT);
bswap(ui32);
```

In Node.js when native code and a recent x86 or ARM processor is available, this
library uses the fastest available SIMD instructions ([PSHUFB (SSSE3) or VPSHUFB
(AVX2)](http://www.felixcloutier.com/x86/PSHUFB.html), [REVn
(NEON)](http://infocenter.arm.com/help/index.jsp?topic=/com.arm.doc.dui0489h/Cihjgdid.html)),
which process multiple array elements simultaneously.

Native code requires one of:
* MSVC 2015 or later
* Clang 3.4.x or later
* GCC 4.8.x or later
* ICC 16 or later

In the browser or when native code is unavailable, this library falls back to
the fastest JavaScript implementation. The JavaScript implementation is also
always explicitly available:

```js
import {js} from "bswap"; // Use javascript implementation explicitly
```

## Benchmarks

Showing millions of elements processed per second when invoked with a
10,000-element array. (Run the benchmark suite to see results for varying array
lengths and other libraries.) Ran on an Intel i7-7700HQ 2.80 GHz processor (AVX2
supported) or Cavium ThunderX 2.0 GHz processor (ARM NEON); Node.js v8.x;
Windows 10 (MSVC) or Ubuntu 16.04 (GCC, Clang). (Note that a 10,000-element
Int16Array fits in L1 cache, whereas a 10,000-element Int32Array or Float64Array
does not.)

| compiler  |    C++ |   JS   | Native:JS | Node.js | Native:Node |
| --------- | -----: | ---:   | --------: | ------: | ----------: |
| **16 bit types (Uint16Array, Int16Array)**
| MSVC 2015 | 32,286 |    625 |     51.7x |  12,141 |        2.7x |
| GCC 8.1   | 31,549 | (same) |     50.5x |   1,507 |       20.9x |
| Clang 6   | 30,238 | (same) |     48.4x |  (same) |       20.1x |
| GCC-ARM   |  2,677 |    183 |     14.6x |     297 |        9.0x |
| **32 bits types (Uint32Array, Int32Array, Float32Array)**
| MSVC 2015 | 12,558 |    342 |     36.7x |   5,840 |        2.2x |
| GCC 8.1   | 12,074 | (same) |     35.3x |   2,361 |        5.1x |
| Clang 6   | 12,587 | (same) |     36.8x |  (same) |        5.3x |
| GCC-ARM   |    670 |     94 |      7.1x |     249 |        2.7x |
| **64 bit types (Float64Array)**
| MSVC 2015 |  6,841 |    179 |     38.2x |   3,043 |        2.2x |
| GCC 8.1   |  6,528 | (same) |     36.5x |   1,790 |        3.6x |
| Clang 6   |  6,598 | (same) |     36.9x |  (same) |        3.7x |
| GCC-ARM   |    382 |     49 |      7.8x |     213 |        1.8x |

There's an AVX512 implementation that is disabled by default. On the Cascade
Lake CPU that I tested on, it is ~28% faster than the AVX2 version when the data
fit in the L1 cache. However, it is ~10% slower than the AVX2 version when the
data come from L2 and ~15% slower from L3. Under the assumption that this module
is more often used with arrays larger than 32KB, I've thus left it disabled.
Sometime maybe I'll make it select between AVX2 and AVX512 depending on the
array length, but this module has no ability to know if the data is resident in
the L1 cache.

## Comparison to other libraries

| Library | Operand | In-Place | 64-bit Type Support | Browser | Speed (vs. bswap)* |
| --- | --- | --- | --- | --- | --- |
| bswap (this) | TypedArray | yes | yes | yes | 1.00 |
| node [`buffer.swap16/32/64`](https://nodejs.org/api/buffer.html#buffer_buf_swap16) | Buffer | yes | since 6.3.0 | no | 0.05 to 0.38 |
| [network-byte-order](https://github.com/mattcg/network-byte-order) | Number/\[Octet\] | no | no | yes | 0.010 |
| [endian-toggle](https://github.com/substack/endian-toggle) | Buffer | no | yes | no | 0.0056 |

\* Higher is better. For 16-bit types, 10k-element arrays. Range given for
Node.js version reflects Windows vs. Linux benchmark.

* **Node.js' built-in
  [buffer.swap16|32|64](https://nodejs.org/api/buffer.html#buffer_buf_swap16)
  methods** (16/32 since v5.10.0; 64 since 6.3.0). Operates in-place. No browser
  support. Slower except for tiny arrays (where it uses the JS implementation).

  In 6.3.0 I added some optimizations to Node.js' implementation. The
  optimizations are effective on Windows, but GCC does not do the same automatic
  vectorization that MSVC does, nor does Node's default build config enable the
  newer SIMD instructions that this library uses.

  <details><summary>Usage</summary>

  ```js
  > Buffer.from(typedArray.buffer).swap16()
  ```

  </details>

* **[endian-toggle](https://github.com/substack/endian-toggle)**. Simple usage,
  operates on a Node.js Buffer, handles any byte size, returns a new buffer
  (does not operate in-place).

  <details><summary>Usage</summary>

  ```js
  > const x = new Uint16Array([2048])
  > toggle(Buffer.from(x.buffer), x.BYTES_PER_ELEMENT * 8)
  <Buffer d2 04 09 07>
  ```

  </details>

* **[network-byte-order](https://github.com/mattcg/network-byte-order)**.
  Operates on a single value at a time (i.e. needs to be looped to operate on an
  array) and has separate `hton` and `ntoh` methods, which do effectively the
  same thing but have different syntaxes. It can operate on strings, but it
  cannot swap 64-bit types.

  <details><summary>Usage</summary>

  ```js
  // Using hton
  > const b = [];
  > nbo.htons(b, 0, 2048);
  > b
  [8, 0]
  
  // or using ntoh
  > const x = new Uint16Array([2048])
  > nbo.ntohs(new Uint8Array(x.buffer, x.byteOffset, 2), 0)
  8
  > const z = new Uint16Array([8])
  > new Uint8Array(z.buffer, z.byteOffset, 2)
  Uint8Array [ 8, 0 ]
  ```

  </details>


changelog.md
# Changelog

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](http://keepachangelog.com/) and this
project adheres to [Semantic Versioning](http://semver.org/).

(Unreleased)
==================
### Changed
### Added
### Fixed

v2.0.0
==================
### Changed
* Breaking: switch to ES Modules
* Update to nan v2.16.0
### Added
* Support BigInt64Array and BigUint64Array
* Keep a changelog


benchmark/benchmark.js
var Benchmark = require("benchmark");
var bswap = require("../bswap.js");
var chalk = require("chalk");

// From benchmark.js
function formatNumber(number) {
	number = String(number).split('.');
	return number[0].replace(/(?=(?:\d{3})+$)(?!\b)/g, ',') +
	(number[1] ? '.' + number[1] : '');
}

function padLeft(input, size) {
	return Array(size - input.length + 1).join(" ") + input;
}

// Modified from benchmark.js
function formatResult(event, times) {
	var hz = event.hz * times;
	var stats = event.stats;
	var size = stats.sample.length;
	var pm = '\xb1';

	var result = " (array length " + times + ")";

	result += ' x ' + chalk.cyan(formatNumber(hz.toFixed(hz < 100 ? 2 : 0))) + ' ops/sec ' + pm +
		stats.rme.toFixed(2) + '% (' + size + ' run' + (size == 1 ? '' : 's') + ' sampled)';

	return result;
}

var byteSizes = [
	// {label: "16 bit", Ctor: Uint16Array},
	{label: "32 bit", Ctor: Uint32Array},
	// {label: "64 bit", Ctor: Float64Array}
];

var arrayLengths = [10000];

var methods = [
	{label: "native", fn: bswap.native},
	{label: "js", fn: bswap.js}
];

try {
	var nbo = require("network-byte-order");
	var fn = function (typedArray) {
		var j, b;
		var al = typedArray.length;
		switch (typedArray.BYTES_PER_ELEMENT) {
			case 2:
				b = new Uint8Array(al * 2);
				for (j = 0; j < al; j++) {
					nbo.htons(b, j * 2, typedArray[j]);
				}
				return b;

			case 4:
				b = new Uint8Array(al * 4);
				for (j = 0; j < al; j++) {
					nbo.htonl(b, j * 4, typedArray[j]);
				}
				return b;

			case 8:
				return;
		}
	};
	methods.push({label: "nbo", fn: fn});
} catch (ex) {
	console.log("network-byte-order unavailable, skipping");
}

try {
	var toggle = require("endian-toggle");
	var fn = function (typedArray) {
		return toggle(Buffer.from(typedArray.buffer), typedArray.BYTES_PER_ELEMENT * 8);
	};
	methods.push({label: "etoggle", fn: fn});
} catch (ex) {
	console.log("endian-toggle unavailable, skipping");
}

try {
	var fn = function (typedArray) {
		switch (typedArray.BYTES_PER_ELEMENT) {
			case 2: return Buffer.from(typedArray.buffer).swap16();
			case 4: return Buffer.from(typedArray.buffer).swap32();
			case 8: return Buffer.from(typedArray.buffer).swap64();
		}
	};
	methods.push({label: "node", fn: fn});
} catch (ex) {
	console.log("Buffer not available in browser, skipping");
}

for (var s = 0; s < byteSizes.length; s++) {
	var byteSize = byteSizes[s];

	console.log(chalk.blue(byteSize.label));

	console.log(
		padLeft("array size", 10),
		padLeft("Native", 15),
		padLeft("JS", 15),
		padLeft("Native:JS", 10),
		padLeft("node", 15),
		padLeft("nbo-hton", 15),
		padLeft("etoggle", 15)
	);

	for (var al = 0; al < arrayLengths.length; al++) {
		var len = arrayLengths[al];

		var input = new byteSize.Ctor(len);
		for (var i = 0; i < input.length; i++) {
			input[i] = i % 255;
		}

		var results = {};

		for (var m = 0; m < methods.length; m++) {
			var method = methods[m];

			if (byteSize.label === "64 bit" && method.label === "nbo") continue;

			new Benchmark.Suite()
				.add({
					name: method.label,
					fn: function () {
						%PrepareFunctionForOptimization(method.fn);
						method.fn(input);
						%OptimizeFunctionOnNextCall(method.fn);
					}
				})
				.on("cycle", function (event) {
					results[method.label] = event.target.hz * len;
				})
				.run();
		}

		console.log(
			padLeft(String(len), 10),
			padLeft(formatNumber(results.native.toFixed(0)), 15),
			padLeft(formatNumber(results.js.toFixed(0)), 15),
			padLeft((results.native / results.js).toFixed(2), 10),
			padLeft("node" in results ? formatNumber(results.node.toFixed(0)) : "", 15),
			padLeft("nbo" in results ? formatNumber(results.nbo.toFixed(0)) : "", 15),
			padLeft("etoggle" in results ? formatNumber(results.etoggle.toFixed(0)) : "", 15)
		);
	}

	console.log();
}
